File: LintCode/java/Longest Common Substring.java
public class Solution {
    /**
     * @param A, B: Two string.
     * @return: the length of the longest common substring.
     */
    public int longestCommonSubstring(String A, String B) {
        int la = A.length();
        int lb = B.length();
        int ans = 0;
        dp = new int[la + 5][lb + 5];
        
        for (int i = 1; i <= la; i++) {
            for (int j = 1; j <= lb; j++) {
                if (A.charAt(i - 1) == B.charAt(j - 1)) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - 1][j - 1] + 1);
                }
                ans = Math.max(ans, dp[i][j]);
            }
        }
        return ans;
    }
    
    private int[][] dp;
}

File: LintCode/cpp/Longest Common Substring.cc
class Solution {
public:    
    /**
     * @param A, B: Two string.
     * @return: the length of the longest common substring.
     */
    int longestCommonSubstring(string &A, string &B) {
        int n = A.size();
        int m = B.size();
        
        init(n, m);
        
        int ans = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (A[i] == B[j]) {
                    dp[i][j] = 
                        (i - 1 >= 0 && j - 1 >= 0? dp[i - 1][j - 1]: 0) + 1;
                }
                ans = max(ans, dp[i][j]);
            }
        }
        return ans;
    }
private:
    void init(int n, int m) {
        dp.resize(n);
        for (auto& _dp: dp) {
            _dp.resize(m);
            fill(_dp.begin(), _dp.end(), 0);
        }
    }
private:
    vector<vector<int> > dp;
};

